core/src/HIP/Kokkos_HIP_Space.cpp

```diff
@@ -74,7 +74,8 @@ hipStream_t get_deep_copy_stream() {
 }  // namespace
 
 void DeepCopyHIP(void* dst, void const* src, size_t n) {
-  KOKKOS_IMPL_HIP_SAFE_CALL(hipMemcpy(dst, src, n, hipMemcpyDefault));
+  hipStream_t s = get_deep_copy_stream();
+  KOKKOS_IMPL_HIP_SAFE_CALL(hipMemcpyAsync(dst, src, n, hipMemcpyDefault, s));
```

Comment:
This is the point of this pull request. We are pretty consistently fencing this variant already so that there is no need to use the (mostly) synchronous variant.
What this pull request does is to follow this policy more thoroughly and thus relax the implicit fencing behavior of the `DeepCopy` without execution space. I updated the pull request description.